### PR TITLE
chore: Enhance pairing test

### LINKED ISSUE #16302: Complete pairings tests
- split test between functional and technical
- use RNG util
- ensure all tests in [doc](https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/docs/proposals/TSS-Library/TSS-Library-Test-Plan.md) are implemented

### PR DESCRIPTION
**Pull request by timo0**
Fix https://github.com/hashgraph/hedera-services/issues/16302

### CODE REVIEW

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128BilinearPairingTest.java [hunk 1]
```diff
@@ -22,38 +22,92 @@
...
+        FieldElement a = fq.fromLong(1); // Identity scalar
+        FieldElement b = fq.fromLong(0); // Zero scalar
+        GroupElement P = g1.generator();
+        GroupElement Q = g2.generator();
+
+        // Test with edge case values
+        testBilinearityProperties(a, b, P, Q);
+    }
+
+    private void testBilinearityProperties(FieldElement a, FieldElement b, GroupElement P, GroupElement Q) {
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128BilinearPairingTest.java]: final

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java [hunk 2]
```diff
@@ -139,4 +140,83 @@ void equalityAndHashCode(final Random rng) {
...
+    void fieldElementMultiplicationZero(final Random rng) {
+        var field = new AltBn128Field();
+        var a = field.random(rng);
+        assertEquals(field.zero(), a.multiply(field.zero())); // a * 0 = 0
+    }
+
+    @Test
+    void createFieldElementFromLargeBigInteger(final Random rng) {
+        var field = new AltBn128Field();
+        var largeValue = BigInteger.valueOf(Long.MAX_VALUE).add(BigInteger.valueOf(1000));
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java]: this is not very large, since field elements are 254 bit numbers

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128Test.java [hunk 3]
```diff
@@ -46,4 +47,99 @@ void fullCircuit(final Random rng) {
...
+    }
+
+    @Test
+    void testHashToCurveConsistency(final Random rng) {
+        final AltBn128 curve = new AltBn128();
+        final Group group1 = curve.group1();
+
+        // Generate random message
+        final byte[] msg = new byte[1024];
+        rng.nextBytes(msg);
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128Test.java:86]: this would ideally be a single call in the SeededRandom

**Comment by mxtartaglia-sl** [comment]
@timo0 , like in hedera-services, let's use conventional commits https://gist.github.com/qoomon/5dfcdf8eec66a051ecd85625518cfd13 names in the PRs.

@rbarkerSL @andrewb1269hg, guys, should something fail if we don't?

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128BilinearPairingTest.java [hunk 4]
```diff
@@ -22,38 +22,92 @@
...
+    @Test
+    void testBilinearityWithEdgeValues() {
+        AltBn128Group g1 = new AltBn128Group(AltBN128CurveGroup.GROUP1);
+        AltBn128Group g2 = new AltBn128Group(AltBN128CurveGroup.GROUP2);
+        AltBn128Field fq = new AltBn128Field();
+
+        FieldElement a = fq.fromLong(1); // Identity scalar
+        FieldElement b = fq.fromLong(0); // Zero scalar
+        GroupElement P = g1.generator();
+        GroupElement Q = g2.generator();
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128BilinearPairingTest.java]: Although it is a valuable test; it is incomplete as an edge scenarios input.

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128BilinearPairingTest.java [hunk 5]
```diff
@@ -22,38 +22,92 @@
...
+        AltBn128Group g2 = new AltBn128Group(AltBN128CurveGroup.GROUP2);
+        AltBn128Field fq = new AltBn128Field();
+
+        GroupElement G1 = g1.generator();
+        GroupElement G2 = g2.generator();
+
+        // Pair generators directly
+        AltBn128BilinearPairing pairing = new AltBn128BilinearPairing(G1, G2);
+
+        assertNotNull(pairing, "Pairing of generators should not be null!");
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128BilinearPairingTest.java]: Missing match check. Please validate externally and add the check. Also it would be helpful to include a link or reference to the tool used to check the billeality is correct.

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java [hunk 6]
```diff
@@ -144,6 +144,120 @@ void batchMultiply(AltBN128CurveGroup gr) {
...
+
+    @ParameterizedTest
+    @EnumSource(AltBN128CurveGroup.class)
+    void testSpecificSerializationPattern(AltBN128CurveGroup gr) {
+        var group = new AltBn128Group(gr);
+        byte[] serialized = group.generator().toBytes();
+
+        // Validate the serialized pattern (for example, against a known hash or specific expected bytes)
+        assertNotNull(serialized);
+        // Further checks as needed, possibly involving expected byte sequences.
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java]: Incomplete test? or wrong test name?

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java [hunk 7]
```diff
@@ -139,4 +140,83 @@ void equalityAndHashCode(final Random rng) {
             assertEquals(elements.get(i).hashCode(), element.hashCode());
         }
     }
+
+    @Test
+    void fieldElementAdditionCommutativity(final Random rng) {
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java]: this test is repeated

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java [hunk 8]
```diff
@@ -139,4 +140,83 @@ void equalityAndHashCode(final Random rng) {
...
+    void fieldElementMultiplicationAssociativity(final Random rng) {
+        var field = new AltBn128Field();
+        var a = field.random(rng);
+        var b = field.random(rng);
+        var c = field.random(rng);
+        assertEquals(a.multiply(b).multiply(c), a.multiply(b.multiply(c)));
+    }
+
+    @Test
+    void fieldElementMultiplicationDistributivity(final Random rng) {
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java]: same test is repeated in fieldEleemnts test

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java [hunk 9]
```diff
@@ -139,4 +140,83 @@ void equalityAndHashCode(final Random rng) {
...
+    @Test
+    void fieldElementAdditionCommutativity(final Random rng) {
+        var field = new AltBn128Field();
+        var a = field.random(rng);
+        var b = field.random(rng);
+        assertEquals(a.add(b), b.add(a));
+    }
+
+    @Test
+    void fieldElementAdditionAssociativity(final Random rng) {
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java]: repeated

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java [hunk 10]
```diff
@@ -139,4 +140,83 @@ void equalityAndHashCode(final Random rng) {
...
+    @Test
+    void fieldElementAdditionInverse(final Random rng) {
+        var field = new AltBn128Field();
+        var a = field.random(rng);
+        var inverseA = field.fromLong(0).subtract(a); // Assuming this would give -a
+        assertEquals(field.zero(), a.add(inverseA)); // a + (-a) = 0
+    }
+
+    @Test
+    void fieldElementMultiplicationCommutativity(final Random rng) {
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java]: repeated in field element tests

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java [hunk 11]
```diff
@@ -139,4 +140,83 @@ void equalityAndHashCode(final Random rng) {
...
+        var field = new AltBn128Field();
+        var a = field.random(rng);
+        var b = field.random(rng);
+        assertEquals(a.multiply(b), b.multiply(a));
+    }
+
+    @Test
+    void fieldElementMultiplicationAssociativity(final Random rng) {
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java]: repeated

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java [hunk 12]
```diff
@@ -139,4 +140,83 @@ void equalityAndHashCode(final Random rng) {
...
+        assertEquals(field.zero(), a.multiply(field.zero())); // a * 0 = 0
+    }
+
+    @Test
+    void createFieldElementFromLargeBigInteger(final Random rng) {
+        var field = new AltBn128Field();
+        var largeValue = BigInteger.valueOf(Long.MAX_VALUE).add(BigInteger.valueOf(1000));
+        assertNotNull(field.fromBigInteger(largeValue)); // Should reduce it within the field
+    }
 }
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128FieldTest.java]: Besides repeating many tests, it seems that they belong to FieldElements test instead.

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java [hunk 13]
```diff
@@ -144,6 +144,120 @@ void batchMultiply(AltBN128CurveGroup gr) {
...
+        // Check that 2 * G = G + G
+        assertEquals(G.multiply(field.fromLong(2)), G.add(G));
+    }
+
+    @ParameterizedTest
+    @EnumSource(AltBN128CurveGroup.class)
+    void testScalarMultiplicationWithLargeValue(AltBN128CurveGroup gr) {
+        var group = new AltBn128Group(gr);
+        var field = new AltBn128Field();
+        var largeScalar = field.fromLong(Long.MAX_VALUE); // Using maximum possible scalar value
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java]: not maximum scalar value

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java [hunk 14]
```diff
@@ -144,6 +144,120 @@ void batchMultiply(AltBN128CurveGroup gr) {
...
+        var G = group.generator();
+
+        // Perform multiplication with the large scalar and check result (e.g., does not throw, is consistent)
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java]: Incomplete? or unnecessary comment?

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java [hunk 15]
```diff
@@ -144,6 +144,120 @@ void batchMultiply(AltBN128CurveGroup gr) {
...
+    @ParameterizedTest
+    @EnumSource(AltBN128CurveGroup.class)
+    void testRandomPointsWithinBounds(AltBN128CurveGroup gr, Random rng) {
+        var group = new AltBn128Group(gr);
+        byte[] seed = new byte[group.seedSize()];
+        rng.nextBytes(seed);
+
+        var randomPoint = group.random(seed);
+
+        // Check that the point is valid (e.g., within the curve bounds if applicable)
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java]: The code could benefit from extra checks.
There is an exception we throw if the point is not on the curve

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupTest.java [hunk 16]
```diff
@@ -121,4 +121,69 @@ void batchOps(AltBN128CurveGroup gr) {
...
+    @Test
+    void hashToCurveConsistency(final Random rng) {
+        var group = new AltBn128Group(AltBN128CurveGroup.GROUP2);
+        final byte[] message = new byte[1024];
+        rng.nextBytes(message);
+
+        var element1 = group.hashToCurve(message);
+        var element2 = group.hashToCurve(message);
+
+        assertEquals(element1, element2); // Consistent mapping of input to output
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupTest.java]: Why is element1 and element2 expected to be different?
I mean I could see the point of sending a different byte array with the same content but calling the method twice on the same input when there is no random source; it just seems to be checking Java determinism.

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java [hunk 17]
```diff
@@ -144,6 +144,120 @@ void batchMultiply(AltBN128CurveGroup gr) {
...
+    @ParameterizedTest
+    @EnumSource(AltBN128CurveGroup.class)
+    void testAdditionAssociativity(AltBN128CurveGroup gr) {
+        var group = new AltBn128Group(gr);
+        var G1 = group.generator();
+        var G2 = group.generator().multiply(new AltBn128Field().fromLong(2));
+        var G3 = group.generator().multiply(new AltBn128Field().fromLong(3));
+
+        // Check (G1 + G2) + G3 = G1 + (G2 + G3)
+        assertEquals(G1.add(G2).add(G3), G1.add(G2.add(G3)));
```
Comment (mxtartaglia-sl) [inline cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128GroupElementTest.java]: Maybe multiply by random scalars instead

**Review by mxtartaglia-sl — CHANGES_REQUESTED** [review]
Timo, there are repeated tests, and some tests seem incomplete or have a wrong comments on top of the asserts.

When we were defining test cases you had the idea of generating points and results with an external library and using those results to validate in Java tests, I don't see any test of that sort and they would be useful.

Also, to match what we've been doing, please add final on all local variables.

**Comment by andrewb1269** [comment]
> @timo0 , like in hedera-services, let's use conventional commits https://gist.github.com/qoomon/5dfcdf8eec66a051ecd85625518cfd13 names in the PRs.
> 
> @rbarkerSL @andrewb1269hg, guys, should something fail if we don't?

Good question. I'd think this is a larger tooling discussion. I know we've used `ci: ` at the beginning of commits, which is not called out in the conventional commits link you have.

Currently I lean towards putting a link in the new employee startup guide, but I don't know if it makes sense to invest tooling time in implementing and enforcing those standards.

@rbarkerSL thoughts?

File: cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128Test.java (hunk 3 above)
Comment (timo0) [reply cryptography/hedera-cryptography-altbn128/src/test/java/com/hedera/cryptography/altbn128/AltBn128Test.java:86]: Currently we inject `Random` not `SeededRandom` maybe we create an issue to change this?

**Review by mxtartaglia-sl — DISMISSED** [review]

**Review by mxtartaglia-sl — APPROVED** [review]
